File: Part 2/cpp_solution/order_tests.cpp
#define BOOST_TEST_DYN_LINK
#define BOOST_TEST_MODULE order_tests
#include <boost/test/unit_test.hpp>
#include "order.h"

BOOST_AUTO_TEST_CASE( test_parse_missing_price )
{
	BOOST_REQUIRE_THROW(ae::order::parse("A:AUDUSD:100"), std::runtime_error);
}

BOOST_AUTO_TEST_CASE( test_invalid_price )
{
	BOOST_REQUIRE_THROW(ae::order::parse("A:AUDUSD:100:blah"), std::runtime_error);
}

BOOST_AUTO_TEST_CASE( test_zero_price_rejected )
{
	BOOST_REQUIRE_THROW(ae::order::parse("A:AUDUSD:100:0"), std::runtime_error);
}

BOOST_AUTO_TEST_CASE( test_invalid_quantity )
{
	BOOST_REQUIRE_THROW(ae::order::parse("A:AUDUSD:xx:23.45"), std::runtime_error);
}

BOOST_AUTO_TEST_CASE( test_zero_quantity )
{
	BOOST_REQUIRE_THROW(ae::order::parse("A:AUDUSD:0:23.45"), std::runtime_error);
}

BOOST_AUTO_TEST_CASE( test_zero_price )
{
	BOOST_REQUIRE_THROW(ae::order::parse("A:AUDUSD:1000:0"), std::runtime_error);
}

BOOST_AUTO_TEST_CASE( test_parse_buy )
{
	auto order = ae::order::parse("A:AUDUSD:100:1.47");
    BOOST_REQUIRE_EQUAL("A", order.participant());
    BOOST_REQUIRE_EQUAL("AUDUSD", order.instrument());
    BOOST_REQUIRE_EQUAL(100, order.quantity());
    BOOST_REQUIRE_EQUAL(1.47, order.price());
    BOOST_REQUIRE(order.is_buy());
}

BOOST_AUTO_TEST_CASE( test_parse_sell )
{
	auto order = ae::order::parse("A:AUDUSD:-100:1.47");
    BOOST_REQUIRE_EQUAL("A", order.participant());
    BOOST_REQUIRE_EQUAL("AUDUSD", order.instrument());
    BOOST_REQUIRE_EQUAL(100, order.quantity());
    BOOST_REQUIRE_EQUAL(1.47, order.price());
    BOOST_REQUIRE(!order.is_buy());
}

File: Part 2/swift_solution/Tests/ExchangeLibTests/OrderTests.swift
import XCTest
@testable import ExchangeLib

class OrderTests : XCTestCase
{
    func testStringInit() throws
    {
        let order = try Order(fromString: "A:AUDUSD:100:1.47")!

        XCTAssertEqual(order.participant, "A")
        XCTAssertEqual(order.instrument, "AUDUSD")
        XCTAssertEqual(order.quantity, 100)
        XCTAssertEqual(order.price, 1.47)
    }

    func testEmptyInit() throws
    {
        XCTAssertNil(try Order(fromString: ""))
    }

    func testInvalidInit() throws
    {
        XCTAssertNil(try Order(fromString: "blah"))
        XCTAssertNil(try Order(fromString: "A:B:C:D"))
        XCTAssertNil(try Order(fromString: "A:B:1.2345:1.0"))
    }

    func testWhitespaceInit() throws
    {
        let order = try Order(fromString: "  A : AUDUSD   : 100  : 1.47   ")!

        XCTAssertEqual(order.participant, "A")
        XCTAssertEqual(order.instrument, "AUDUSD")
        XCTAssertEqual(order.quantity, 100)
        XCTAssertEqual(order.price, 1.47)
    }
   
}
